text/0000-regex-1.0.md

```diff
@@ -0,0 +1,980 @@
+or there exists at least one widely used third party implementation. It
+therefore seems reasonable for Rust to do something similar.
+
+The `regex` crate specifically serves many use cases, most of which are somehow
+related to searching strings for patterns. Describing regular expressions in
+detail is beyond the scope of this RFC, but briefly, these core use cases are
+supported in the main API:
+
+1. Testing whether a pattern matches some text.
+2. Finding the location of a match of a pattern in some text.
+3. Finding the location of a match of a pattern---and locations of all its
+   capturing groups---in some text.
+4. Iterating over successive non-overlapping matches of (2) and (3).
+
+The expected outcome is that the `regex` crate should be the preferred default
+choice for matching regular expressions when writing Rust code. This is already
+true today; this RFC formalizes it.
+
+# Detailed design
+[design]: #detailed-design
+
+## Syntax
+[syntax]: #syntax
+
+### Evolution
+[evolution]: #evolution
+
+The public API of a `regex` library *includes* the syntax of a regular
+expression. A change in the semantics of the syntax can cause otherwise working
+programs to break, yet, we'd still like the option to expand the syntax if
+necessary. Thus, this RFC proposes:
+
+1. Any change that causes a previously invalid regex pattern to become valid is
+   *not* a breaking change. For example, the escape sequence `\y` is not a
+   valid pattern, but could become one in a future release without a major
+   version bump.
+2. Any change that causes a previously valid regex pattern to become invalid
+   *is* a breaking change.
+3. Any change that causes a valid regex pattern to change its matching
+   semantics *is* a breaking change. (For example, changing `\b` from "word
+   boundary assertion" to "backspace character.")
+
+Bug fixes are exceptions to both (2) and (3).
+
+Another interesting exception to (2) is that compiling a regex can fail if the
+entire compiled object would exceed some pre-defined user configurable size.
+In particular, future changes to the compiler could cause certain instructions
+to use more memory, or indeed, the representation of the compiled regex could
+change completely. This could cause a regex that fit under the size limit to
+no longer fit, and therefore fail to compile. These cases are expected to be
+extremely rare in practice. Notably, the default size limit is `10MB`.
+
+### Concrete syntax
+[concrete-syntax]: #concrete-syntax
+
+The syntax is exhaustively documented in the current public API documentation:
+http://doc.rust-lang.org/regex/regex/index.html#syntax
+
+To my knowledge, the evolution as proposed in this RFC has been followed since
+`regex` was created. The syntax has largely remain unchanged with few
+additions.
+
+### Expansion concerns
+[expansion-concerns]: #expansion-concerns
+
+There are a few possible avenues for expansion, and we take measures to make
```

Comment:
Another possibility here is "compatibility" flags, e.g. if there is a breaking change for whatever reason, people can opt-in to it with some sort of flag like `(?X)`, ala `(?i)`.